### DISCUSSION TITLE
Update only specific dependency

### DISCUSSION THREAD

**Discussion by wrbbz** [renovatebot/renovate#20523] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

32.31.1

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### If you're self-hosting Renovate, tell us what version of the platform you run.

15.7.7 EE

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

I would like to tell renovate to update only specific dependency. Dynamically.
The use case is:
I have a repo with a lot of submodules and when one submodule (the repo itself) updates, I would like to update only that submodule with renovate instead of updating all of them.

P.S. I know, that I am running not the newest version. I have got some troubles with its configuration. Thus, I am using the 32

### What you tried so far.

I have tried to use `--include-paths`, but as I understood, it only works with package names and not with dependency names:

Renovate config:
```
{
  "git-submodules": {
     "enabled": true
  },
  "packageRules": [
    {
      "description": "Teams Documentation update",
      "matchPaths": ["content/**", ".gitmodules"],
      "automerge": true,
      "platformAutomerge": true
    }
  ]
}
```

I have tried the following run:

```bash
renovate --include-paths 'content/submoduleA','.gitmodules' path/to/repo
```

Still, renovate is trying to update all submodules

### Relevant debug logs

<details><summary>Logs</summary>

```
...
DEBUG: Found 1 package file(s) (repository=path/to/repo)
 INFO: Dependency extraction complete (repository=path/to/repo)
       "baseBranch": "main",
       "stats": {
         "managers": {"git-submodules": {"fileCount": 1, "depCount": 2}},
         "total": {"fileCount": 1, "depCount": 2}
       }
...
DEBUG: 2 flattened updates found: content/submoduleA, content/submoduleB (repository=path/to/repo)
...
DEBUG: Updated 1 package files (repository=path/to/repo, branch=renovate/content-submoduleA-digest)
DEBUG: Updated 1 lock files (repository=path/to/repo, branch=renovate/content-submoduleA-digest)
       "updatedArtifacts": ["content/submoduleA"]
...
DEBUG: Updated 1 package files (repository=path/to/repo, branch=renovate/content-submoduleB-digest)
DEBUG: Updated 1 lock files (repository=path/to/repo, branch=renovate/content-submoduleB-digest)
       "updatedArtifacts": ["content/submoduleB"]
...
```

</details>

**Comment by viceice** [comment, marked as answer]
by default renovate is enabled for all deps, so you need to disable them manually and exclude  the one you like to update

**Reply by wrbbz** [reply]
Ok. I'll try this approach. Can I exclude them dynamically one by one? Instead of excluding them in config file

**Reply by viceice** [reply]
what do you mean with dynamic?

**Reply by wrbbz** [reply]
On different jobs in GitLab CI it will be different dependencies (submodules). On jobA execution I want to update only submoduleA, on jobB I want to update submoduleB etc. Thus, I can not put these excludes in `renovate.json` file as is. And I am not sure that environment variables could be parsed in `renovate.json`

**Reply by viceice** [reply]
you can't have different renovate jobs for a single repo. if you try that, you get into trouble.

**Reply by wrbbz** [reply]
So I can not update different submodules on request?

**Reply by viceice** [reply]
you can, use dependency dashboard approval[^1]

[^1]: https://docs.renovatebot.com/key-concepts/dashboard/#dependency-dashboard-approval-workflow

**Reply by wrbbz** [reply]
As I understand correctly, it will be a bottleneck on approvals. For example, we have `submoduleA` and `submoduleB` projects that are connected as submodules to `upstream-project`. `upstream-project` has a rennovate.json config. And `renovate-runner` project has configured CI job that updates `upstream-project`. 
What I want:
On update of `submoduleA` create a MR in `upstream-project` only for `submoduleA` and do not check any other submodules. And after that automerge the created MR so developers do not have to wait for my (or someone else) approve and merge of MR.

As I understand, I have two solutions right now:
1. Run renovate on all submodules and update all of them 
2. Use dependency dashboard approval with manual approve of every submodule

Is it right or am I missing something?

**Reply by rarkins** [reply]
I'm not sure I understand your question, but MR creation, MR approval, and automerge are all completely configurable in a granular manner using packageRules. You can exclude submoduleA from approvals and set it to automerge, for example

**Reply by wrbbz** [reply]
No. The question is not about configuration of MR approvals. It is more about an ability to specify which dependency to update
Maybe it is better to abstract away from my point of view. Maybe you can advise another solution for my problem.

So, the problem is:
I've got few repositories that are connected to the upstream one as a submodules:
```mermaid
flowchart LR
submoduleA-->Upstream
submoduleB-->Upstream
...
submoduleZ-->Upstream
```
With update of submoduleA I want to update **only** submoduleA ref in upstream repository. On update of submoduleB - only submoduleB ref, etc.
In perfect conditions - I would like this updates without any bottlenecks. Update of submodule leads to update of its ref in default branch

**Reply by rarkins** [reply]
How is this different to how things behave today?

e.g. if submoduleA is the only updated dependency, Renovate will update *only* submoduleA in your upstream. If later submoduleB is updated, you'll get an update for only submoduleB. By default, Renovate updates dependencies individually.

**Reply by wrbbz** [reply]
Yeah. Thats true. My point was about step-by-step update but after all thoughts I figured out that it is a complicated and unnecessary behavior. Thank you for your time! 

I'll mark that thread as an answer
